Add extra_hosts to yml configuration --add-hosts

chernjie opened the pull request:
Add extra_hosts to yml configuration

Followup from:
- #597
- #754 
- #848 
- #958 

### Example:
```yml
extra_hosts:
 - "docker:162.242.195.82"
 - "fig:50.31.209.229"
```

Signed-off-by: CJ <lim@chernjie.com>

chernjie commented:
@tdesvenain @sampwing I have signed your commits after rebasing from the latest master.

tdesvenain commented:
Great, thanks !

chernjie commented:
I am wondering if it is possible to add an alias to `extra_hosts`, e.g. `hosts`

pedrokiefer commented:
Shouldn't it be possible to add and extra_host that is another container, but not actually linked to the current container.

Something like:
```
web:
  build: .
worker:
  build: .
  extra_hosts:
     - web
```

aanand commented:
@pedrokiefer Why not just use a link? You'll get the `/etc/hosts` entry.

chernjie commented:
@pedrokiefer : What @aanand said, use `link` for containers, use `extra_hosts` for non-containers

pedrokiefer commented:
I had some kind of circular dependency between containers, so I thought ```extra_hosts``` would solve that to me.

chernjie commented:
rebased with CI fixes

chernjie commented:
@aanand I would appreciate your feedback/input with this PR. The integration tests are working well. Are there anything that I have missed?

aanand commented on compose/service.py:
```diff
@@ -616,3 +620,23 @@ def split_port(port):
+    if extra_hosts_config is None:
+        return None
+
+    if isinstance(extra_hosts_config, basestring):
```
`basestring` should be `six.string_types()` for Python 3 compatibility

aanand commented:
Nice. Only thing is that I think supporting the list-of-dicts format is really strange. I would rather support:

- list of `host:ip` strings
- (optionally) dictionary of `host` string to `ip` string

This would be consistent with `environment` and `labels`.

chernjie commented:
@aanand I do not have a strong opinion on this.
The slight difference between `list-of-dicts` vs `list-of-strings` is the spacing immediately after the colon `:`. These can easily be introduced from mistakes.

```yaml
dicts:
 docker: 162.242.195.82
 fig: 50.31.209.229
list-of-dicts:
 - docker: 162.242.195.82
 - fig: 50.31.209.229
list-of-strings:
 - docker:162.242.195.82
 - fig:50.31.209.229
```

That said, the current `build_extra_hosts` code supports all three.

@tdesvenain, do you have any input since you were the one who wrote the code?

aanand commented:
I agree that the syntax is error-prone (I blame YAML for being too permissive), but it feels like a hack to allow lists of dicts and we don't do it anywhere else.

Perhaps in the examples for list-of-strings, we should always quote them:

```yaml
list-of-strings:
 - "docker:162.242.195.82"
 - "fig:50.31.209.229"
```

chernjie commented:
@aanand updated the examples

tdesvenain commented:
Hi,
I'm afraid that many people who use to work with yaml will make the same
mistake,
In my humble opinion if we use the ':' character in docker-compose syntax,
we'd better use it the common way in yaml, as key/value separator.
On the top of that, we actually describe a mapping (host/ip), so dictionary
seems more accurate than string.

But actually if this introduces inconsistency, i'm ok with removing this.
There,
- we have to validate the format to alert user (before my commit we just
got an exception that looked like an issue)
- I agree that we should at least quote the whole string in documentation

2015-04-04 16:26 GMT+02:00 chernjie <notifications@github.com>:

> @aanand <https://github.com/aanand> updated the examples
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/compose/pull/1158#issuecomment-89586085>.
>



-- 
Thomas Desvenain

Téléphone : 09 51 37 35 18

tdesvenain commented:
After having further reflexion about it, i think my commits could be
reverted as they introduce too much inconsistency against other declarations

2015-04-04 19:01 GMT+02:00 Thomas Desvenain <thomas.desvenain@gmail.com>:

> Hi,
> I'm afraid that many people who use to work with yaml will make the same
> mistake,
> In my humble opinion if we use the ':' character in docker-compose syntax,
> we'd better use it the common way in yaml, as key/value separator.
> On the top of that, we actually describe a mapping (host/ip), so
> dictionary seems more accurate than string.
>
> But actually if this introduces inconsistency, i'm ok with removing this.
> There,
> - we have to validate the format to alert user (before my commit we just
> got an exception that looked like an issue)
> - I agree that we should at least quote the whole string in documentation
>
> 2015-04-04 16:26 GMT+02:00 chernjie <notifications@github.com>:
>
>> @aanand <https://github.com/aanand> updated the examples
>>
>> --
>> Reply to this email directly or view it on GitHub
>> <https://github.com/docker/compose/pull/1158#issuecomment-89586085>.
>>
>
>
>
> --
> Thomas Desvenain
>
> Téléphone : 09 51 37 35 18
>



-- 
Thomas Desvenain

Téléphone : 09 51 37 35 18

chernjie commented:
@aanand @tdesvenain I have explicitly removed support for `list-of-dicts`, added some new test to make sure that only list and mappings are supported.

aanand commented on compose/service.py:
```diff
@@ -619,3 +623,32 @@ def split_port(port):
+    if isinstance(extra_hosts_config, list):
+        extra_hosts_dict = {}
+        for extra_hosts_line in extra_hosts_config:
+            if isinstance(extra_hosts_line, dict):
```
I think it'd be better to check `if not isinstance(extra_hosts_line, six.string_types)` and raise an error saying `extra_hosts_config must be either a list of strings or a string->string mapping`. We shouldn't need to do anything special with the list-of-dicts case.

chernjie replied:
Updated code

tdesvenain replied:
Where the code is updated ?
Did you fix the unit tests i wrote which rely with dict format ?

2015-04-14 16:01 GMT+02:00 chernjie <notifications@github.com>:

> In compose/service.py
> <https://github.com/docker/compose/pull/1158#discussion_r28328697>:
>
> > @@ -619,3 +623,32 @@ def split_port(port):
> >
> >      external_ip, external_port, internal_port = parts
> >      return internal_port, (external_ip, external_port or None)
> > +
> > +
> > +def build_extra_hosts(extra_hosts_config):
> > +    if not extra_hosts_config:
> > +        return {}
> > +
> > +    if isinstance(extra_hosts_config, list):
> > +        extra_hosts_dict = {}
> > +        for extra_hosts_line in extra_hosts_config:
> > +            if isinstance(extra_hosts_line, dict):
>
> Updated code
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/compose/pull/1158/files#r28328697>.
>



-- 
Thomas Desvenain

Téléphone : 09 51 37 35 18

aanand commented on docs/yml.md:
````diff
@@ -87,6 +87,23 @@ external_links:
+extra_hosts:
+ - "docker:162.242.195.82"
+ - "fig:50.31.209.229"
+```
````
"docker" and "fig" are strange host names - why would I have a host called "fig"? (cc @moxiegirl - any best practices here?)

Could you add an example with a dict? Best if the strings are quoted, e.g.

```
extra_hosts:
  somehost: "162.242.195.82"
  otherhost: "50.31.209.229"
```

chernjie replied:
Updated examples

chernjie commented:
hmmm... @aanand Jenkins stopped running a few days ago:
https://jenkins.dockerproject.com/job/Compose-PRs/
![image](https://cloud.githubusercontent.com/assets/275729/7138723/9f8abbe8-e2f4-11e4-9a9a-3244db0806c6.png)

aanand commented:
Gave it a kick and it's up again. Looks like you need to wrap your lists in `set(...)` when asserting equality so it ignores ordering.

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "addhosts" git@github.com:chernjie/compose.git somewhere
$ cd somewhere
$ git rebase -i HEAD~14
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

chernjie commented:
Added `.sort` instead of `set`, would this work?
cc: @aanand

chernjie commented:
@aanand any updates to this PR? Any feedback would be appreciated

aanand commented:
`set(...)` is preferable for consistency with other tests.

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "addhosts" git@github.com:chernjie/compose.git somewhere
$ cd somewhere
$ git rebase -i HEAD~15
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

chernjie commented:
@aanand Updated with `set(...)`

aanand commented:
Fantastic. Could you squash and rebase? Thanks.

chernjie commented:
@aanand I have rebased and squashed all my commits. I have left the other author's contribution as individual commit so as not to lose their contribution information. I hope that's okay

dnephin commented on compose/config.py:
```diff
@@ -41,6 +42,7 @@
 
 DOCKER_CONFIG_HINTS = {
     'cpu_share': 'cpu_shares',
+    'add_host': 'extra_hosts',
```
maybe `'extra_host': 'extra_hosts',` as well ?

chernjie replied:
Added @ 86a08c0

dnephin commented on compose/service.py:
```diff
@@ -619,3 +623,28 @@ def split_port(port):
+        extra_hosts_config = extra_hosts_dict
+
+    if isinstance(extra_hosts_config, dict):
+        return extra_hosts_config
```
It might be a good idea to only support one of these formats.  We've supported both in the past, but I'm not sure that's necessary going forward. It seems to cause confusion and extra code.

If we decide to keep lists (or both), I would suggest `<host>=<ip>` (equals instead of colon) syntax instead. If you forget the quotes it confuses yaml into things its a mapping, and that can be very counter-intuitive to users.  If we use an equals, it should understand that it is a string, not a mapping.

If we decide to keep mappings (or both), we should probably do some basic validation of the keys and values, maybe even just assert that they are Truthy (not None or empty strings).

dnephin replied:
Oh, I just realized this discussion has already happened. I'll defer to whatever @aanand decides now that I'm registered my concerns.

dnephin commented:
I like this, I think it addresses the "external" type described in #988

LGTM

aanand commented:
LGTM

rysiekpl commented:
Oh this is sweet, and I'm kind of craving this feature right now. Any date set for 1.3.0 release?
